### DISCUSSION TITLE
WLALinker in other folders

### DISCUSSION THREAD

**Discussion by Pokeman2003** [vhelin/wla-dx#487] [Q&A] [answered]
Quick addendum:For those that need a quickly concocted solution while waiting for new binaries to release, it turns out that it doesn't do a relative path based on the wlalinker itself, it does a relative path based on the directory the cmd.exe that called upon it is currently looking. A very easy fix would to just be cd into the directory with the listfile and objects, and then call wlalinker from there.

Hi all! I posted to SMSPower about this question earlier, but the only one so far to offer help, maxim, doesn't know how to fix it either. Long story short, I have a batch file automating my project compilation. It looks a bit like this.
```batch
del "finloc"\object.o
del "finloc"\listfile
"compilerlocation\wla-z80.exe" -o "projectlocation"\project.asm "finloc"\object.o
echo [objects]>"finloc"\listfile
echo object.o>>"finloc"\listfile
"compilerlocation\wlalinker.exe" -d -r -v -s "finloc"\listfile "finloc"\output.sms
```
Compilerlocation, finloc, and projectlocation are all relative paths, and thus have quotation marks around for me to avoid worrying about a space somewhere.

Now, this code seems like it SHOULD work, but when I run it, the linker returns an error, complaining to me that it could not open `object.o`. The file DOES exist in the intended location, so I'm not entirely sure what's going wrong. I tried putting `"finloc"\object.o` in the listfile, but then it just complains it can't open "finloc", and -L switch did nothing to change my situation.

It was suggested that wlalinker worked with relativity, which would potentially make sense.  So the question is, how do I get WLALinker to operate in other folders that it's not in?

**Comment by vhelin** [comment]
Is there any reason why you don't specify the paths like

```
"projectlocation\project.asm"
"finloc\object.o"
```

Personally I've never seen anybody try to use something like "finloc"\object.o

**Reply by Pokeman2003** [reply]
The reason for me doing this is very simple. Finloc is actually a placeholder since it's supposed to be portable. The reason why I'm not giving the actual path is because in many cases, it could look like `C:\Users\MYUSER\Desktop\SMS\comp`, or `E:\comp`, or even `C:\Users\MY USER\P R O B L E M D I R E C T O R Y\comp`. Actually, `compilerlocation` looks like that, `finloc` would be `\fin\%project%`. The reason why I need the quotation marks is so that I, passing this through a call :function, don't end up with a mess of arguments. So, I pass it as `"%cd%\what I need\"` instead of just `%cd%\what I need`, as the latter would split the path apart as a bunch of arguments automatically thanks to the spaces.

**Comment by vhelin** [comment]
For reference, I tried both in Cygwin under Windows 10 and they seemed to work

![image](https://user-images.githubusercontent.com/6417746/156330228-c67ff618-9ba4-4f2a-b1f3-fbf3cd267a68.png)

I think what happens here is that the command line app processes those quotation marks and just feeds WLA a string without quotation marks. So it's command line dependent how those quotation marks are processed and how they work.

WLALINK's linkfile on the other hand is another case as its has its own parser. That might not support (and if I remember right it doesn't) things like "path"/main.asm. I'll come back to this later today...

**Comment by maxim-zhao** [comment]
I think the explanation is confusing. If I have a working linkfile I can do this:
```
wlalink.exe -d -r -v -S linkfile "ps1jert.sms"
```
If I navigate to another folder (e.g. the parent) and pass relative paths, it does not work:
```
wlalink.exe -d -r -v -S path\to\linkfile "ps1jert.sms"
```
```
LOAD_FILE_DATA: Could not open file "ps1jert.sms.o".
```
The issue is that paths in the linkfile are not interpreted as relative to its location in the file system, instead they are interpreted as relative to the current directory.

**Reply by vhelin** [reply]
Ok well this discussion needs to be moved to the "Issues" side here on Github. :)

Perhaps adding a flag to WLALINK that'd make paths in linkfile to be relative to its location would fix this?

**Reply by maxim-zhao** [reply]
Indeed - look at "create issue from discussion" up above...

**Comment by vhelin** [comment, marked as answer]
If you guys now try the latest sources and give WLALINK flag -R, does it work for you?

**Reply by Pokeman2003** [reply]
I would but CMake requires an installation, and the computer I am doing this on is not my own, nor do I have administrator privileges. I'll try after I get home.
Also, for those that need a quickly concocted solution while waiting for new binaries, it turns out that it doesn't do a relative path based on the wlalinker itself, it does a relative path based on the directory the cmd.exe that called upon it is currently looking, so just cd into the directory you need, and then run wlalinker from there and that'll actually work for some bizarre reason.

**Reply by vhelin** [reply]
FYI, you can get binaries built using the very latest sources from these Azure Pipelines icons in README.md

![image](https://user-images.githubusercontent.com/6417746/156438487-a2c60782-7ddb-43ac-bebe-5670b26c9d84.png)

**Reply by vhelin** [reply]
For example, click on the Windows icon, then under "Related" click on text "1 Published", then you see "drop" and take mouse to far right above "..." and left click on that and "Download artifacts"...

**Reply by Pokeman2003** [reply]
Oh huh, I wasn't aware. Then I'll try that. Thanks man!

**Reply by Pokeman2003** [reply]
Well, it definitely works, but as I said in my issue reply, it retroactively worked on my old binaries too.

**Reply by vhelin** [reply]
I just added that -R flag today to WLALINK...

**Comment by lhsazevedo** [comment]
Nice to have :)

**Reply by vhelin** [reply]
Absolutely. And the more people report these nice-to-have features the more we can implement them and everybody just wins. :)
